sql/core/benchmarks/FilterPushdownBenchmark-results.txt

```diff
@@ -2,669 +2,809 @@
+Native ORC Vectorized                         4828 / 5158          3.3         307.0       1.5X
+Native ORC Vectorized (Pushdown)               524 /  540         30.0          33.3      13.8X
+InMemoryTable Vectorized                      3695 / 3785          4.3         234.9       2.0X
+InMemoryTable Vectorized (Pushdown)           1702 / 1745          9.2         108.2       4.3X
 
-OpenJDK 64-Bit Server VM 1.8.0_181-b13 on Linux 3.10.0-862.3.2.el7.x86_64
-Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
+Java HotSpot(TM) 64-Bit Server VM 1.8.0_191-b12 on Mac OS X 10.12.6
+Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
 InSet -> InFilters (values count: 10, distribution: 90): Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------
-Parquet Vectorized                          10572 / 10590          1.5         672.1       1.0X
-Parquet Vectorized (Pushdown)                  713 /  716         22.1          45.3      14.8X
-Native ORC Vectorized                         6808 / 6815          2.3         432.9       1.6X
-Native ORC Vectorized (Pushdown)               530 /  541         29.7          33.7      19.9X
+Parquet Vectorized                            7303 / 7388          2.2         464.3       1.0X
+Parquet Vectorized (Pushdown)                  471 /  477         33.4          29.9      15.5X
+Native ORC Vectorized                         4848 / 5005          3.2         308.2       1.5X
+Native ORC Vectorized (Pushdown)               518 /  550         30.4          32.9      14.1X
+InMemoryTable Vectorized                      3665 / 3790          4.3         233.0       2.0X
+InMemoryTable Vectorized (Pushdown)           1689 / 1717          9.3         107.4       4.3X
 
-OpenJDK 64-Bit Server VM 1.8.0_181-b13 on Linux 3.10.0-862.3.2.el7.x86_64
-Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
+Java HotSpot(TM) 64-Bit Server VM 1.8.0_191-b12 on Mac OS X 10.12.6
+Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
 InSet -> InFilters (values count: 50, distribution: 10): Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
 ------------------------------------------------------------------------------------------------
-Parquet Vectorized                          10871 / 10882          1.4         691.2       1.0X
-Parquet Vectorized (Pushdown)               11104 / 11110          1.4         706.0       1.0X
-Native ORC Vectorized                         7088 / 7104          2.2         450.7       1.5X
-Native ORC Vectorized (Pushdown)               665 /  677         23.6          42.3      16.3X
+Parquet Vectorized                            7474 / 7577          2.1         475.2       1.0X
+Parquet Vectorized (Pushdown)                 7649 / 7707          2.1         486.3       1.0X
+Native ORC Vectorized                         5167 / 5457          3.0         328.5       1.4X
+Native ORC Vectorized (Pushdown)               637
```

Comment:
Do you know why this is so slow?

Reply:
I don't know yet, but I am very interested in this case.

Reply:
The reason is that Parquet/ORC [set all nullability fields to true](https://github.com/apache/spark/blob/2d085c13b7f715dbff23dd1f81af45ff903d1a79/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L504-L505), but InMemoryTable does't. The `InferFiltersFromConstraints` optimizer will [add `isNotNull` to all columns](https://github.com/apache/spark/blob/d87d30e4fe9c9e91c462351e9f744a830db8d6fc/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala#L93-L95).